### PR TITLE
Add interface classes to object IDs

### PR DESCRIPTION
**Pull request by xbrianh**
These object ids are used for S3 object key construction, as well as internal logic.

These will be used in the refactored version of flash-flood.

### CODE REVIEW

File: flashflood/identifiers.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+class JournalID(str):
+    DELIMITER = "--"
+
+    @classmethod
+    def make(cls, start_timestamp: str, end_timestamp: str, version: str, blob_id: str):
+        return cls(start_timestamp + cls.DELIMITER + end_timestamp + cls.DELIMITER + version + cls.DELIMITER + blob_id)
+
+    @classmethod
+    def from_key(cls, key):
+        return cls(key.rsplit("/", 1)[1])
```
Comment (amarjandu) [inline flashflood/identifiers.py]: should there be restrictions to what form the `key` that is submitted can take?

Comment (xbrianh) [reply flashflood/identifiers.py]: regex patterns could be used to constrain key format. That can be handled in future work.

File: flashflood/identifiers.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    def timestamp(self) -> str:
+        return self._parts()[2]
+
+    @property
+    def action(self) -> JournalUpdateAction:
+        return self._parts()[3]
+
+    @staticmethod
+    def prefix_for_journal(journal_id: JournalID):
+        return journal_id[::-1]
```
Comment (chmreid) [inline flashflood/identifiers.py]: `[::-1]` will reverse the list/string. Why is the prefix for the journal just the journal ID in reverse? I wouldn't call that a prefix. Also why would the "prefix" of an ID be the ID itself but backwards? It means journals never share a prefix. But, it looks like there are other parts of the code using this reverse-string logic (e.g., `reverse_journal_id` above).

A comment in the docstring of the class or the function would be great to clarify what's going on.

File: flashflood/identifiers.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+class JournalUpdateID(str):
+    """
+    This defines the id used to compose the object key on storage for journal updates.
+    """
+    DELIMITER = "--"
+
+    @classmethod
+    def make(cls, journal_id: str, event_id: str, action: JournalUpdateAction):
+        reverse_journal_id = journal_id[::-1]
```
Comment (chmreid) [inline flashflood/identifiers.py]: Should this use the `prefix_for_journal()` function instead?

File: flashflood/identifiers.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        return cls(reverse_journal_id + cls.DELIMITER
+                   + event_id + cls.DELIMITER
+                   + timestamp_now() + cls.DELIMITER
+                   + action.name)
+
+    @classmethod
+    def from_key(cls, key):
```
Comment (chmreid) [inline flashflood/identifiers.py]: Nit: suggest renaming this function to `make_from_key()`

File: flashflood/identifiers.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+class JournalID(str):
+    DELIMITER = "--"
+
+    @classmethod
+    def make(cls, start_timestamp: str, end_timestamp: str, version: str, blob_id: str):
+        return cls(start_timestamp + cls.DELIMITER + end_timestamp + cls.DELIMITER + version + cls.DELIMITER + blob_id)
+
+    @classmethod
+    def from_key(cls, key):
```
Comment (chmreid) [inline flashflood/identifiers.py]: Nit: would suggest renaming this function to `make_from_key()`

File: flashflood/identifiers.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+TOMBSTONE_SUFFIX = ".dead"
+
+
+class JournalUpdateAction(Enum):
+    UPDATE = "update"
+    DELETE = "delete"
+
+
+class JournalID(str):
```
Comment (chmreid) [inline flashflood/identifiers.py]: Should add a docstring here that briefly explains what a journal ID is/what it is used for/what form it takes.

```
A Journal ID is a string that identifies journal events by their start/end date, version, and blobs affected. (...)
```

Maybe this is clear to you from the name and from reading the code, but it's a little confusing to me because there are Journal updates, actions, events, IDs, update IDs, etc., some of which are represented by classes, and it's hard to parse out how these are related or unrelated.

File: flashflood/identifiers.py (hunk 3 above)
Comment (chmreid) [reply flashflood/identifiers.py]: (Also, if so, should rename the variable from `reverse_journal_id` to `journal_prefix`

File: flashflood/identifiers.py (hunk 2 above)
Comment (xbrianh) [reply flashflood/identifiers.py]: Update markers are written as S3 objects with the id `{reverse_journal_id}{event_id}{action}`. To list updates for a specific journal you need to use the prefix `{reverse_journal_id}`. This is done to increase the randomness of the leading characters of update keys, since non-reversed journal ids ends with a UUID.

I'll address your readability concerns in a future PR.

File: flashflood/identifiers.py (hunk 3 above)
Comment (xbrianh) [reply flashflood/identifiers.py]: Probably. Will fix.

File: flashflood/identifiers.py (hunk 4 above)
Comment (xbrianh) [reply flashflood/identifiers.py]: IMO that does not increase readability.
